src/execution_plan/execution_plan_modify.c

```diff
@@ -176,9 +176,16 @@ void ExecutionPlan_PushBelow(OpBase *a, OpBase *b) {
 }
 
 void ExecutionPlan_NewRoot(OpBase *old_root, OpBase *new_root) {
-	/* child is the current root operation and parent is a new operation. */
-	assert(!old_root->parent && !new_root->parent && !new_root->children);
-	_OpBase_AddChild(new_root, old_root);
+	/* The new root should have no parent, but may have children if we've constructed
+	 * a chain of traversals/scans. */
+	assert(!old_root->parent && !new_root->parent);
+
+	// Find the deepest child of the new root operation.
+	OpBase *tail = new_root;
+	while(tail->childCount > 0) tail = tail->children[0];
```

Comment:
This is safe given the context in which I've introduced it, but you're right that it could pose future issues if used more broadly.

The reason it's safe right now is that the new root only has child operations if it's a traversal chain (see the [associated assertion change](https://github.com/RedisGraph/RedisGraph/pull/721/files#diff-569479632b0ecc7b1b7a2ce44601357bL180).) By nature, this chain doesn't branch - we make a different call if our traversals are part of a Cartesian product.

For the moment, I don't think I want to add too much unnecessary complexity here, but I will comment in this area more comprehensively!